### PR TITLE
cache: adds unmanaged groups to be cached and loaded in the identity

### PR DESCRIPTION
**Pull request by TLGINO**
* closes https://github.com/inveniosoftware/invenio-app-rdm/issues/2186

### CODE REVIEW

**Comment by jrcastro2** [comment]
tests passing locally

File: invenio_communities/utils.py
Comment (jrcastro2) [inline invenio_communities/utils.py:75]: This is not ideal, but there is no way to fetch concrete keys from flask-caching. That's why we are accessing protected methods.

File: invenio_communities/cache/cache.py
Comment (ntarocco) [inline invenio_communities/cache/cache.py:29]: maybe a good idea to add `pass` or `raise NotImplementedError`? or maybe mark them as `abstract`?

File: invenio_communities/cache/cache.py
Comment (ntarocco) [inline invenio_communities/cache/cache.py:14]:
```suggestion
class IdentityCache:
```

File: invenio_communities/cache/redis.py
Comment (ntarocco) [inline invenio_communities/cache/redis.py:22]:
```suggestion
        super().__init__(app=app)
```

File: invenio_communities/cache/redis.py
Comment (ntarocco) [inline invenio_communities/cache/redis.py:26]: Is it, in the end, necessary to bypass the cache abstraction and go directly to Redis? Is it because we discussed separating in different cache databases, so that when it is cleaned, only that db is touched?

File: invenio_communities/ext.py
Comment (ntarocco) [inline invenio_communities/ext.py:15]: `six` is not needed any more :) Please remove.

File: invenio_communities/members/services/components.py
Comment (ntarocco) [inline invenio_communities/members/services/components.py:58]: is `type` mandatory? Or optional?

File: invenio_communities/members/services/components.py
Comment (ntarocco) [inline invenio_communities/members/services/components.py:63]: Here there is room for improvement:
- we get the user id
- we wrap it around an Identity
- then, when deleting, we get again the `id`

File: invenio_communities/members/services/schemas.py
Comment (ntarocco) [inline invenio_communities/members/services/schemas.py:143]: is the name optional or mandatory in the DB?

File: invenio_communities/utils.py
Comment (ntarocco) [inline invenio_communities/utils.py:30]:
```suggestion
    # This cache has a risk of being eternal therefore a cronjob that clears all the cache has to be set on a daily basis
```

File: invenio_communities/members/services/components.py
Comment (jrcastro2) [reply invenio_communities/members/services/components.py:58]: It`s mandatory, will fix it, thanks!

File: invenio_communities/members/services/schemas.py
Comment (jrcastro2) [reply invenio_communities/members/services/schemas.py:143]: Name is optional [see here](https://github.com/inveniosoftware/invenio-accounts/blob/master/invenio_accounts/models.py#L69)

File: invenio_communities/alembic/37b21951084c_update_role_id_type.py
Comment (alejandromumo) [inline invenio_communities/alembic/37b21951084c_update_role_id_type.py:19]: question: where is this revision coming from? I can't find it in invenio org.

File: invenio_communities/alembic/37b21951084c_update_role_id_type.py
Comment (alejandromumo) [inline invenio_communities/alembic/37b21951084c_update_role_id_type.py:17]: question: this revision is from [invenio-requests](https://github.com/inveniosoftware/invenio-requests/blob/18594a369a0c568b2c83a14ed9e1cb3fef447ae8/invenio_requests/alembic/a14fa442680f_create_tables.py#L15). Shouldn't it be `depends_on=a14fa442680f`?

File: invenio_communities/cache/redis.py
Comment (alejandromumo) [inline invenio_communities/cache/redis.py:45]: note: some keys might need to be deleted maybe? Passing `None` would be enough?

File: invenio_communities/utils.py
Comment (alejandromumo) [inline invenio_communities/utils.py:22]: shelve: we're adding some data fetching logic (from the cache) in `utils`, does not feel right. However it can be tackled afterwards.

File: invenio_communities/members/services/components.py
Comment (alejandromumo) [inline invenio_communities/members/services/components.py:32]: major: `community` must be defined. Otherwise it can raise an exception. 


question: `member.group_id` wiped before calling this component?

File: tests/cache/test_identity_redis_cache.py
Comment (alejandromumo) [inline tests/cache/test_identity_redis_cache.py:16]: Idea: we can take inspiration for more tests from [here](https://github.com/inveniosoftware/invenio-rdm-migrator/blob/master/tests/streams/test_cache.py) 😇

File: tests/conftest.py
Comment (alejandromumo) [inline tests/conftest.py:61]: Maybe we can test the expires (e.g. small timeouts)

File: invenio_communities/members/services/service.py
Comment (alejandromumo) [inline invenio_communities/members/services/service.py:245]: question: where did this code go?

**Review by alejandromumo — CHANGES_REQUESTED** [review]
left multiple questions

**Review by ntarocco — COMMENTED** [review]
Can be merged after discussed changes and alembic.

File: invenio_communities/alembic/37b21951084c_update_role_id_type.py
Comment (jrcastro2) [reply invenio_communities/alembic/37b21951084c_update_role_id_type.py]: Fixed, was a leftover

File: invenio_communities/config.py
Comment (jrcastro2) [reply invenio_communities/config.py:302]: 3 is used because 0, 1 and 2 are already being used. Since this is a config variable it is configurable already.

File: invenio_communities/cache/redis.py
Comment (jrcastro2) [reply invenio_communities/cache/redis.py:45]: By passing `None` there is no timeout and the keys are not deleted. you can set the value in seconds and after the set time they will be deleted. Otherwise, if we need to delete a key we can use the delete func below.

File: invenio_communities/members/services/service.py
Comment (jrcastro2) [reply invenio_communities/members/services/service.py:245]: This was a hack to map the name to the id. Not any more needed. Now you can have a role with an ID which is different from the name.

File: invenio_communities/utils.py
Comment (jrcastro2) [reply invenio_communities/utils.py:87]: Yes, we can remove it :+1:

File: invenio_communities/utils.py
Comment (jrcastro2) [reply invenio_communities/utils.py:22]: Not sure why `utils` os not right. More than fetching we are setting some values in the cache but every time we set we need to fetch them to add the new element to the list as the set func from the cache does not allow to "append". I guess we could add an `add` func in the RedisCache class that does this logic (but it would assume that the value of the key is a list/set).

File: invenio_communities/members/services/components.py
Comment (jrcastro2) [reply invenio_communities/members/services/components.py:32]: This function takes care of updating the members, for both `user` and `groups`. However, for users we do not need to pass the community but for groups we need. If we try to upgrade a group wihout passing the community it has to raise an exception. Alternatively we can split the functions to differentiate the change of membership between `users` and `groups`, but if we try to update a group without passing a community it will always raise an exception as it has to.

Comment (jrcastro2) [reply invenio_communities/members/services/components.py:32]: > question: member.group_id wiped before calling this component?

As far as I have tested it's not wiped before calling the component. Why do you think it's being wiped before?

File: tests/conftest.py
Comment (jrcastro2) [reply tests/conftest.py:61]: Right! Good point! thanks :)

File: tests/cache/test_identity_redis_cache.py
Comment (jrcastro2) [reply tests/cache/test_identity_redis_cache.py:16]: I have improved the tests with your suggestion of testing the default timeout expiration. I think all cases are now covered unless I am missing something

File: tests/members/test_members_services.py
Comment (jrcastro2) [reply tests/members/test_members_services.py:131]: We've updated the fixture that creates the tests to use the new type (String), as long as the uuids are converted to strings it's OK, I think it's enough if the tests check the type of the id, independently of how the id was generated (as long as it's a string), unless I am missing something

File: invenio_communities/alembic/fbe746957cfc_create_member_tables.py
Comment (jrcastro2) [inline invenio_communities/alembic/fbe746957cfc_create_member_tables.py:19]: MAJOR: Question for reviewers, with this change we are able to use the upgrade command from alembic even from scratch, and I think it's correct that this dependency directly depends on the recipe that creates. (that revision id could potentially be moved to `depends_on`, to be tested if )

Comment (slint) [reply invenio_communities/alembic/fbe746957cfc_create_member_tables.py:19]: To me this sounds like the "alembic merge" is the way to enforce the order we were discussing IRL. We should do that at least for the recipes requiring this explicitness, i.e. PK type changes...

Now for the rest, my gut feeling is that we should also fix them in the long term, but it looks like that's a much bigger  task... Do we know roughly how many places that is? Is there a way to incrementally fix them across modules?

Comment (jrcastro2) [reply invenio_communities/alembic/fbe746957cfc_create_member_tables.py:19]: To me the "alembic merge"  is the solution, only if we need to downgrade to that recipe, or to force the path through that recipe, the `depends_on` does not enforce that path but just tells  that it needs the revision_id to be processed before being executed (meaning that many others recipes could be executed in between both recipes).

About how many places should we fix it ... So far it looks like it works with this one, however there might be other places that are affected. To find them, we should review all the recipes, to fix them incrementally, we can just fix them as they start to fail :sweat_smile:

**Review by ppanero — COMMENTED** [review]
copyright years headers 😇

File: invenio_communities/alembic/fbe746957cfc_create_member_tables.py
Comment (kpsherva) [reply invenio_communities/alembic/fbe746957cfc_create_member_tables.py:19]: discussed in person, decided to go ahead with the solution and mitigate any consequences it might cause

File: invenio_communities/members/services/components.py
Comment (kpsherva) [reply invenio_communities/members/services/components.py:32]: I agree with Manuel, you have set the community to be None as default, maybe we need to raise an exception to check if it is passed? or in the conditional statement we need to check if community is not None?

File: invenio_communities/utils.py
Comment (kpsherva) [reply invenio_communities/utils.py:22]: I agree, it does not feel right, because I wouldn't expect from utility function to be obliged to have application context (and current_identities_cache kinda indicates that), but I can see it was done like this previously...
In the end you are checking for duplicates in the cache, why not making the .set method of the cache class to check it ?

File: invenio_communities/utils.py
Comment (kpsherva) [reply invenio_communities/utils.py:71]: agree

File: invenio_communities/utils.py
Comment (jrcastro2) [reply invenio_communities/utils.py:122]: Updating roles is only related to name or description changes of the role which will have no impact on it's members, nor permissions to access a community

File: invenio_communities/utils.py
Comment (jrcastro2) [reply invenio_communities/utils.py]: The cache get does not allow to return a default if missing

**Pull request merged**
